ProjectorSum object for TFQ

tonybruguier opened the pull request:
This is a follow-up on #4331 which introduced ProjectorString objects. The present PR aims at adding ProjectorSum objects.

tonybruguier commented:
I would suggest @MichaelBroughton as a reviewer, but I am happy to receive help from anyone.

MichaelBroughton reviewed (changes requested):
This is a good first pass, there are some broader changes we should look into before doing a detailed review:
1. Probably want to move this to `/ops/linear_combinations/py` with PauliSum and the linearcombination of gates.
2. For the `__mul__`, `__imul__` etc, you should follow along a little more closely to what is done is PauliSum. A good starting point is in the in-place operations shouldn't need to make a copy and slow things down.
3. You might want to add a few more useful user facing methods like `__iter__`, `__len__` etc. which will allow you to tighten up other areas of the code.

tonybruguier commented:
Thanks, @MichaelBroughton 

* I moved the files to the location you suggested
* I removed some of the copying of the objections, inspiring me from `PauliSum`. Some copying is still present, but it is also present in `PauliSum`, so maybe it's OK
* I added the user-facing methods you suggested. I also took the liberty of adding `__bool__` too.

PTAL when you can?

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,183 @@ def __format__(self, format_spec: str) -> str:
+    def __init__(
+        self, linear_dict: Optional[value.LinearDict[FrozenSet[Tuple[raw_types.Qid, int]]]] = None
+    ):
+        """Contructor for ProjectorString
```
"Constructor for ProjectorSum"

tonybruguier replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,183 @@ def __format__(self, format_spec: str) -> str:
+
+        Args:
+            projector_qids: Ordered collection of qubits that determine the subspace
+                in which the matrix representation of the ProjectorString is to
```
ProjectorSum

tonybruguier replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,183 @@ def __format__(self, format_spec: str) -> str:
+        state_vector: np.ndarray,
+        qid_map: Mapping[raw_types.Qid, int],
+    ) -> float:
+        """Expectation of the sum projection from a state vector.
```
Nit: Compute the expectation value of this ProjectorSum given a state vector.

tonybruguier replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,183 @@ def __format__(self, format_spec: str) -> str:
+        state: np.ndarray,
+        qid_map: Mapping[raw_types.Qid, int],
+    ) -> float:
+        """Expectation of the sum of projections from a density matrix.
```
Nit: Compute the expectation of this ProjectorSum given a state vector.

tonybruguier replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,183 @@ def __format__(self, format_spec: str) -> str:
+    def __bool__(self) -> bool:
+        return bool(self._linear_dict)
+
+    def __iadd__(self, other: 'ProjectorSum'):
+        self._linear_dict += other._linear_dict
+        return self
+
+    def __add__(self, other: 'ProjectorSum'):
+        result = self.copy()
+        result += other
+        return result
```
Looking at how `__add__` and `__iadd__` are implemented in PauliSum:

```python3
def __iadd__(self, other):
      if isinstance(other, numbers.Complex):
          other = PauliSum.from_pauli_strings([PauliString(coefficient=other)])
      elif isinstance(other, PauliString):
          other = PauliSum.from_pauli_strings([other])
    
      if not isinstance(other, PauliSum):
          return NotImplemented
    
      self._linear_dict += other._linear_dict
      return self
    
def __add__(self, other):
  if not isinstance(other, (numbers.Complex, PauliString, PauliSum)):
      if hasattr(other, 'gate') and isinstance(other.gate, identity.IdentityGate):
          other = PauliString(other)
      else:
          return NotImplemented
  result = self.copy()
  result += other
  return result
```
I think this can still be tightened up to type check for ProjectorString, ProjectorSum, complex numbers and return `NotImplemented` otherwise. I think a similar approach would be worth adopting for `sub`, `mul` etc.

tonybruguier replied:
I tried to do the exact same thing as for PauliSum as much as possible.

I also took the liberty of adding more tests, hope that's OK.

MichaelBroughton reviewed (changes requested):
Getting closer. I think we still need some tigther math operations for the `ProjectorSum` class. In addition to this I would split up the tests to be a little more targetted for each math operation. Instead of just having `test_projector_sum_operations` have a few more targetted tests like PauliSum does: i.e. `test_paulisum_mul_paulistring` and `test_paulisum_mul_paulisum` just for multiplication functionality.

tonybruguier commented:
Thanks, @MichaelBroughton 

I tried to follow the code as much as possible. I also took the liberty of adding more tests. In particular, since I was confused earlier (and still) about whether or not to call `.copy()`, I decided to add tests to make sure the original objects were unmodified. I can, of course, remove these tests.

PTAL

tonybruguier commented:
@MichaelBroughton 
Now with the ability to add ProjectorString objects to ProjectorSum objects as you had suggested, I think.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations_test.py:
```diff
@@ -1662,3 +1662,253 @@ def test_expectation_from_density_matrix_two_qubit_states():
         )
+
+
+def test_projector_sum_expectations():
```
Can we split this into two tests, one for state vector and one for density matrix. Also maybe using `cirq.testing.random_supoerposition` again to add a little omre complexity here ?

tonybruguier replied:
I decided to split in 3, instead of 2: get the matrix, expectation from state vector, expectation from density matrix.

Also, I decided to add coefficients to both projectors.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,210 @@ def __format__(self, format_spec: str) -> str:
+
+        Args:
+            state_vector: An array representing a valid state vector.
+            qubit_map: A map from all qubits used in this ProjectorString to the
```
Nit: ProjectorSum

tonybruguier replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/linear_combinations.py:
```diff
@@ -729,3 +732,210 @@ def __format__(self, format_spec: str) -> str:
+
+        Args:
+            state: An array representing a valid  density matrix.
+            qubit_map: A map from all qubits used in this ProjectorString to the
```
Nit: ProjectorSum

tonybruguier replied:
Done.

MichaelBroughton reviewed (commented):
This is looking pretty good. Just a few minor nits and expanding some of the tests a little bit then we should be good to merge.

tonybruguier commented:
Thanks, @MichaelBroughton . PTAL when you have the time?

MichaelBroughton reviewed (approved):
LGTM
